File: benchmarks/Java/src/cd/Vector3D.java
package cd;


public final class Vector3D {
  public final double x;
  public final double y;
  public final double z;

  public Vector3D(final double x, final double y, final double z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  public Vector3D plus(final Vector3D other) {
    return new Vector3D(x + other.x,
                        y + other.y,
                        z + other.z);
  }

  public Vector3D minus(final Vector3D other) {
    return new Vector3D(x - other.x,
                        y - other.y,
                        z - other.z);
  }

  public double dot(final Vector3D other) {
    return x * other.x + y * other.y + z * other.z;
  }

  public double squaredMagnitude() {
    return this.dot(this);
  }

  public double magnitude() {
    return Math.sqrt(squaredMagnitude());
  }

  public Vector3D times(final double amount) {
    return new Vector3D(x * amount,
                        y * amount,
                        z * amount);
  }
}

File: benchmarks/C++HJ/cd/Vector3D.cpp
#include "Vector3D.h"
#include <iostream>

namespace CD {
        
    Vector3D::Vector3D(double x, double y, double z) {
        _x = x;
        _y = y;
        _z = z;
    }

    shared_ptr<Vector3D> Vector3D::plus(shared_ptr<Vector3D> other) const {
        return make_shared<Vector3D>(_x + other->_x, _y + other->_y, _z + other->_z);
    }

    shared_ptr<Vector3D> Vector3D::minus(shared_ptr<Vector3D> other) const {
        return make_shared<Vector3D>(_x - other->_x, _y - other->_y, _z - other->_z);
    }

    double Vector3D::dot(shared_ptr<Vector3D> other) {
        return _x * other->_x + _y * other->_y + _z * other->_z;
    }

    double Vector3D::squaredMagnitude() {
        return dot(shared_from_this());
    }


    double Vector3D::magnitude() {
        return sqrt(squaredMagnitude());
    }

    shared_ptr<Vector3D> Vector3D::times(double amount) const {
        return make_shared<Vector3D>(_x * amount, _y * amount, _z * amount);
    }
};